### PR TITLE
Add software update code

### PR DESCRIPTION
**Pull request by Taylorr82**

### CODE REVIEW

File: services/source/service_utilities.c [hunk 1]
```diff
@@ -0,0 +1,151 @@
...
+    cnv.cnv8[2] = from[2];
+    cnv.cnv8[1] = from[1];
+    cnv.cnv8[0] = from[0];
+    *to = cnv.cnv32;
+}
+
+void cnv8_16LE(uint8_t *from, uint16_t *to) {
+    union _cnv cnv;
+
+    cnv.cnv8[0] = from[1];
```
Comment (rcunrau) [inline services/source/service_utilities.c:79]: All of these conversions are very curious, since memcpy (or casting) could have been used for almost all of them. This one does a swap, which assume the host is running big-endian. What if the host is actually little-endian already?

File: bl_eeprom.c [hunk 2]
```diff
@@ -44,53 +45,68 @@ char eeprom_get_boot_type() {
     char bootType;
     TI_Fee_ReadSync(BOOT_TYPE_BLOCK, BOOT_TYPE_OFFSET, (uint8_t *)(&bootType), BOOT_TYPE_LEN);
     TI_FeeJobResultType res = TI_Fee_GetJobResult(0);
+    TI_Fee_ErrorCodeType error = TI_FeeErrorCode(0);
```
Comment (rcunrau) [inline bl_eeprom.c]: Why add this if you're not doing anything with the result?

File: bl_flash.c [hunk 3]
```diff
@@ -166,20 +166,20 @@ uint32_t Fapi_BlockProgram( uint32_t Bank, uint32_t Flash_Address, uint32_t Data
...
 		 (void)Fapi_enableAutoEccCalculation();
 		 (void)Fapi_setActiveFlashBank((Fapi_FlashBankType)Bank);
-	     (void)Fapi_enableMainBankSectors(0xFFFF);                    /* used for API 2.01*/
+	     (void)Fapi_enableMainBankSectors(0xFFFF);
 	}else {
          return (1);
-	}
-
+	}*/
+	(void)Fapi_setActiveFlashBank((Fapi_FlashBankType)Bank);
```
Comment (rcunrau) [inline bl_flash.c]: Hmm, is this fixing a bug? It's hard to believe this change has anything to do with restartable updates - perhaps it should be in a separate PR?

File: bl_flash.c [hunk 4]
```diff
@@ -166,20 +166,20 @@ uint32_t Fapi_BlockProgram( uint32_t Bank, uint32_t Flash_Address, uint32_t Data
...
 	while( FAPI_CHECK_FSM_READY_BUSY != Fapi_Status_FsmReady );
-	while( FAPI_GET_FSM_STATUS != Fapi_Status_Success );
+	//while( FAPI_GET_FSM_STATUS != Fapi_Status_Success );
 
     while( SizeInBytes > 0)
 	{
-		Fapi_issueProgrammingCommand((uint32_t *)dst,
+        Fapi_StatusType success = Fapi_issueProgrammingCommand((uint32_t *)dst,
```
Comment (rcunrau) [inline bl_flash.c:187]: Why get the result and then not look at it, unless the argumentless macros expect a variable named success (shudder)

File: ex2_bootloader.dil [hunk 5]
```diff
@@ -1,4 +1,4 @@
-# TMS570LC4357ZWT 01/15/22 22:06:50
+# TMS570LC4357ZWT 01/31/22 21:38:07
```
Comment (rcunrau) [inline ex2_bootloader.dil:1]: This file looks like it has been auto-generated. Is it supposed to be checked in?

File: main/bl_main.c [hunk 6]
```diff
@@ -121,6 +124,8 @@ char get_boot_type(int rstsrc, boot_info *b_inf) {
 }
 
 void bl_main(resetSource_t rstsrc) {
+    RAISE_PRIVILEGE;
+    RESET_PRIVILEGE;
```
Comment (rcunrau) [inline main/bl_main.c]: This looks like a NO-OP

File: services/source/updater.c [hunk 7]
```diff
@@ -0,0 +1,355 @@
...
+#include <csp/csp_endian.h>
+#include <main/system.h>
+#include <os_task.h>
+#include "service_utilities.h"
+#include <FreeRTOS-Plus-CLI/FreeRTOS_CLI.h>
+#include "printf.h"
+#include "privileged_functions.h"
+
+static struct __attribute__((packed)) {
+    uint32_t start_address;
```
Comment (rcunrau) [inline services/source/updater.c]: I usually use void* or uintptr_t for addresses, in case their size ever changes

File: services/source/updater.c [hunk 8]
```diff
@@ -0,0 +1,355 @@
...
+                    2,
+                    /* Store the parameter string length. */
+                    &addr_str_len);
+
+    if (addr_str_len > 10) {
+        snprintf(pcWriteBuffer, xWriteBufferLen, "Hex string too long\n");
+        return pdFALSE;
+    }
+
+    const char *hex_values = addr_str + 2;
```
Comment (rcunrau) [inline services/source/updater.c]: I assume this is for "0x" - do you know for sure it's there?

File: services/source/updater.c [hunk 9]
```diff
@@ -0,0 +1,355 @@
...
+ * @detail
+ *      This function will try to malloc smaller and smaller buffers until it succeeds
+ *      In this case, having any buffer at all is more important than size
+ * @param buf
+ *      Pointer to the pointer to malloc the buffer in to
+ * @return
+ *      size of buffer that was allocated for
+ */
+void *get_buffer(int32_t size) {
+    return pvPortMalloc(size);
```
Comment (rcunrau) [inline services/source/updater.c:176]: The function doesn't really seem to be doing what the comments say it is doing

File: services/source/updater.c [hunk 10]
```diff
@@ -0,0 +1,355 @@
...
+        // wait for connection, timeout
+        if ((conn = csp_accept(sock, DELAY_WAIT_TIMEOUT)) == NULL) {
+            /* timeout */
+            continue;
+        }
+
+        while ((packet = csp_read(conn, 50)) != NULL) {
+            if (updater_app(packet) != SATR_OK) {
+                // something went wrong, this shouldn't happen
+                csp_buffer_free(packet);
```
Comment (rcunrau) [inline services/source/updater.c:345]: If the csp_send is returning a response, don't you want to return both happy and sad responses?

File: bl_flash.c (hunk 3 above)
Comment (Taylorr82) [reply bl_flash.c]: The flash bank the code interacts with must always be active and after a reboot it won't be active, so resuming after a reboot relies on this. There's no harm in setting it every time so it's very simple to just set it

File: bl_flash.c (hunk 4 above)
Comment (Taylorr82) [reply bl_flash.c:187]: I should probably check the value

File: bl_eeprom.c (hunk 2 above)
Comment (Taylorr82) [reply bl_eeprom.c]: This code isn't supposed to be here, I deleted it. Have to investigate

File: services/source/service_utilities.c (hunk 1 above)
Comment (Taylorr82) [reply services/source/service_utilities.c:79]: This is code taken from the main project, we should review it there

File: bl_eeprom.c (hunk 2 above)
Comment (Taylorr82) [reply bl_eeprom.c]: Looks like you commented on an old commit here

File: ex2_bootloader.dil (hunk 5 above)
Comment (Taylorr82) [reply ex2_bootloader.dil:1]: This needs to be checked in because it's the hardware configuration file for halcogen

File: main/bl_main.c (hunk 6 above)
Comment (Taylorr82) [reply main/bl_main.c]: yes

File: services/source/updater.c (hunk 7 above)
Comment (Taylorr82) [reply services/source/updater.c]: the flashapi library expects uint32 so to silence warnings (and errors if the size changes) I kept it as uint32.

File: services/source/updater.c (hunk 8 above)
Comment (Taylorr82) [reply services/source/updater.c]: Since it's just a debug function I didn't pay too much attention to error handling. However, hex2int does do this with error checking now, so I'll remove it

File: services/source/updater.c (hunk 9 above)
Comment (Taylorr82) [reply services/source/updater.c:176]: no, it doesn't. Whoops

File: services/source/updater.c (hunk 10 above)
Comment (Taylorr82) [reply services/source/updater.c:345]: I don't know what you mean
